### ISSUE TITLE
atexit is also unusable on AIX

### ISSUE THREAD

**Issue by hyc** [openssl/openssl#23575]
The atexit option added for #23135 is apparently also needed on AIX. This was reported to us in OpenLDAP https://bugs.openldap.org/show_bug.cgi?id=10176

Really, using atexit() in a shared library is simply wrong, and every non-glibc system and every glibc system older than 2.2.3 will do the wrong thing with it. You should instead be telling the linker to invoke your cleanup function as a destructor. You should not be relying on such a nonstandard behavior of atexit in the first place. https://pubs.opengroup.org/onlinepubs/9699919799/functions/atexit.html

**Comment by nhorman** [comment]
A destructor would be a good solution, but openssl conforms to C89, and so their use is not realy feasible.  Instead we are planning on eliminating the use of atexit in a future release, and mandating applications call OPENSSL_cleanup on their own prior to exit.  Currently this fix:
https://github.com/openssl/openssl/pull/23394
In the master branch allows you to suppress the use of atexit() at build time, which should give you what you want.  Enabling it will require your application call OPENSSL_cleanup manually prior to exit,

**Comment by hyc** [comment]
Looking at this again, the original problem that was reported to us was https://bugs.openldap.org/show_bug.cgi?id=9952

We called SSL_CTX_free() in libldap's destructor, which crashed because OpenSSL's atexit() handler had already run. We "fixed" this by changing libldap to also use atexit(), so that our SSL_CTX_free would happen before OpenSSL's atexit handler. And that change caused our ITS#10176. It seems there's nothing we can safely change in libldap to address this. At this point we are going to revert our own use of atexit. The bug in 9952 will still exist; either we remove our call to SSL_CTX_free(), and accept the memory leak, or libssl is changed to not dereference the NULL pointers that are left behind by its cleanup, to avoid the crash (see the stack trace in ITS#9952).

**Comment by davidben** [comment]
> mandating applications call OPENSSL_cleanup on their own prior to exit

What exactly does OpenSSL do in atexit that requires this? There is no need to free things like memory allocations or even file descriptors immediately before exit because the process is imminently shutting down anyway. I could imagine an application wanting to flush some buffered I/O before exit, but a cryptography library needing this is bizarre.

**Comment by kroeckx** [comment]
As far as I know, the only reason to do this is to avoid tools like valgrind complaining about a memory leak. Having openssl clean up its things helps avoid false positives in other applications.

**Comment by hlandau** [comment]
Does that mean if we drop the atexit call, we could argue it isn't a breaking change?

We should be calling OPENSSL_cleanup in our own unit tests for valgrind etc.

**Comment by mattcaswell** [comment]
> Does that mean if we drop the atexit call, we could argue it isn't a breaking change?

I don't think we can argue that its not a breaking change. As soon as we drop this we will get a flood of reports about "memory leaks".

**Comment by davidben** [comment]
A correct leak detector should incorporate reachability into how it reports things. We primarily use LSan, which gets this right, and this says valgrind does too:
https://developers.redhat.com/blog/2021/04/23/valgrind-memcheck-different-ways-to-lose-your-memory

It sounds then like the reports will be limited to people misunderstanding valgrind's output, and they can be pointed to this document. And you can always tell them to call `OPENSSL_cleanup`, if they are really truly insistent.

It'll be tedious to deal with that confusion, but the status quo is crashes (or perhaps even exploitable memory errors!) in real world applications because OpenSSL's use of `atexit` doesn't match your callers use of threads.

It also sounds like `OPENSSL_cleanup`'s documentation should be fixed. It's *not* a mandate on applications, but an optional call for applications that really want to waste cycles on program shutdown. It probably needs to come with some warnings too. In my experience with OpenSSL callers, it's quite common for library code to call such cleanup functions, not realizing the implications for other libraries in the same address space that use OpenSSL. For BoringSSL, we intentionally make all the global cleanup functions into no-ops because they don't actually do anything useful, and it's so common for folks to get this wrong.

**Comment by alex** [comment]
As a consumer distributing a python extension module linked with OpenSSL, the use of `atexit` is a big footgun for us, as it implicitly ties the lifetimes of the Python process, our extension module, and the OpenSSL library together in a way that makes it very easy to produce UAFs and such (for example, when any thread lives longer than the main thread).

**Comment by kroeckx** [comment]
I think one of the motivations to change to use atexit is that libraries got this wrong all the time, calling it and causing problems for other users.

**Comment by davidben** [comment]
That doesn't really solve the problem, just makes it worse. The right thing for the libraries wasn't `atexit`. As we've seen throughout all this, `atexit` is a *problem* for libraries. The libraries cannot guarantee anything about how the top-level application above uses threads, and the top-level application may not even know OpenSSL is involved because it's abstracted away. The right thing for the libraries is to just let process shutdown free things.

Libraries called `OPENSSL_cleanup` because they saw a function named "cleanup" and assumed they needed to call it. It's better now, but OpenSSL is old, with a long history of very unclear, complicated, and poorly documented initialization requirements. This led to projects just copy-paste a random initialization-looking functions from other code. Then they see a cleanup function and naturally assume they need to call it.

Even OpenSSL's current documentation is still misleading here:

> OPENSSL_INIT_NO_ATEXIT
> By default OpenSSL will attempt to clean itself up when the process exits via an "atexit" handler. Using this option suppresses that behaviour. This means that the application will have to clean up OpenSSL explicitly using OPENSSL_cleanup().

https://www.openssl.org/docs/man3.2/man3/OPENSSL_cleanup.html

As discussed on this bug, this is false. The application does not need to clean OpenSSL up explicitly, and indeed doing so risks security vulnerabilities in multi-threaded programs.

The right behavior is to use `OPENSSL_INIT_NO_ATEXIT` and then _not_ call `OPENSSL_cleanup`. But this is a footgun, so OpenSSL should just never register `atexit` and then fix the documentation to discourage `OPENSSL_cleanup`. (I would go even further. Make `OPENSSL_cleanup` a no-op, because we have to assume applications were written against OpenSSL's old documentation, which incorrect recommended `OPENSSL_cleanup`. The only way to safely meet that documented API contract is to make `OPENSSL_cleanup` do nothing. This is what BoringSSL does.)

**Comment by kroeckx** [comment]
I think the cleanup function was introduced at the same time as the atexit, and there were 1 or more other cleanup functions before that that now do nothing.

**Comment by davidben** [comment]
Ah interesting. Okay, so anyone calling `OPENSSL_cleanup` is probably doing so less out of confusion and more out of the bad documentation. I think the bad documentation is still sufficient to warrant turning this into a no-op, but either way the important point is that libraries can't safely call `atexit`. In order to call `atexit`, you need to know something about how the whole program uses threads (namely whether threads outlive `main`) and only the top-level application can know that.

Which is fine because there is no reason to free reachable globals on process shutdown.

**Comment by beldmit** [comment]
There is such a reason. It is an easy metrics to have 0 leaks at shutdown. If you have some, you have to deal with them - whether they are one-time or not, for example.

**Comment by hyc** [comment]
In the meantime, (again looking at the stack trace in our ITS#9952) would it be acceptable to patch around these crashes? e.g.
````
diff --git a/crypto/threads_pthread.c b/crypto/threads_pthread.c
index ad6c259d92..b6cd0cad23 100644
--- a/crypto/threads_pthread.c
+++ b/crypto/threads_pthread.c
@@ -657,6 +657,8 @@ CRYPTO_RWLOCK *CRYPTO_THREAD_lock_new(void)
 
 __owur int CRYPTO_THREAD_read_lock(CRYPTO_RWLOCK *lock)
 {
+    if (lock == NULL)
+        return 0;
 # ifdef USE_RWLOCK
     if (pthread_rwlock_rdlock(lock) != 0)
         return 0;
@@ -672,6 +674,8 @@ __owur int CRYPTO_THREAD_read_lock(CRYPTO_RWLOCK *lock)
 
 __owur int CRYPTO_THREAD_write_lock(CRYPTO_RWLOCK *lock)
 {
+    if (lock == NULL)
+        return 0;
 # ifdef USE_RWLOCK
     if (pthread_rwlock_wrlock(lock) != 0)
         return 0;
@@ -687,6 +691,8 @@ __owur int CRYPTO_THREAD_write_lock(CRYPTO_RWLOCK *lock)
 
 int CRYPTO_THREAD_unlock(CRYPTO_RWLOCK *lock)
 {
+    if (lock == NULL)
+        return 0;
 # ifdef USE_RWLOCK
     if (pthread_rwlock_unlock(lock) != 0)
         return 0;
````

**Comment by petrovr** [comment]
Kurt Roeckx wrote:
> I think the cleanup function was introduced at the same time as the atexit, and there were 1 or more other cleanup functions before that that now do nothing.

Functionality to initialize and clean library is existing since beginning. Macro or set of functions or single function is just different implementation.


In general clean exit could ensure that resources are freed. It is not about cases related to cryptography library.
It could be free of resources like file descriptors or sockets. Note cryptography library may use such resources indirectly.

Regards,
Roumen

**Comment by kroeckx** [comment]
My suggestion for applications affected by this is to call OPENSSL_init_crypto with OPENSSL_INIT_NO_ATEXIT, and not call the cleanup function.

If it's always a problem on the OS, use the build time option.

In some future version we might change the default not to use atexit.

**Comment by davidben** [comment]
> There is such a reason. It is an easy metrics to have 0 leaks at shutdown. If you have some, you have to deal with them - whether they are one-time or not, for example.

That is easy, but given the number of projects that `atexit` is broken, it's a bit too simplistic. There are much better leak detectors that don't have this problem. Both LeakSanitizer and valgrind do a reachability check. That's the better way to check for leaks.

`atexit` is also worse for performance! Messing with your allocator's bookkeeping for a ton of individual allocations is slower than just letting the kernel tear the whole thing down.

> It could be free of resources like file descriptors or sockets. Note cryptography library may use such resources indirectly

Sockets and file descriptors, like memory, are already destroyed for you on process shutdown.

> My suggestion for applications affected by this is to call OPENSSL_init_crypto with OPENSSL_INIT_NO_ATEXIT, and not call the cleanup function.

While this is the right behavior, it's not what the documentation suggests. Given the problems, it sounds like you all at least should update that.

But also keep in mind (this perhaps should be documented) that this is not a reliable solution for a library. Only the first call gets to determine whether `atexit` happens. If some random other library in your program initializes OpenSSL first and doesn't pass the flag (perhaps the library author never ran into this problem and followed OpenSSL's documentation), you still get `atexit`.

So while it's the right behavior, and the best way for applications to fix it right now, OpenSSL still has something to do here.

**Comment by hyc** [comment]
Just to note - we faced a similar dilemma in OpenLDAP's libldap, and we fixed this 22 years ago.
https://git.openldap.org/openldap/openldap/-/commit/ffa4b2634389ecc129f59a1de05d44f62b6bb871

Use a destructor when the compiler supports it; only use atexit when there's no support for destructors and you're building a static library. It's silly not to use a feature that everybody's compiler supports today, and choosing to rely on a nonstandard glibc behavior instead.

**Comment by rsbeckerca** [comment]
> Kurt Roeckx wrote:
> I think the cleanup function was introduced at the same time as the atexit, and there were 1 or more other cleanup functions before that that now do nothing.
> Functionality to initialize and clean library is existing since beginning. Macro or set of functions or single function is just different implementation. In general clean exit could ensure that resources are freed. It is not about cases related to cryptography library. It could be free of resources like file descriptors or sockets. Note cryptography library may use such resources indirectly. Regards, Roumen

The use of `atexit` changed at 3.3 with new test introductions and changed existing behaviour of OpenSSL regardless of what was there "in the beginning". This broke some platforms as evidence from the number of `atexit` cases here. A potentially reasonable solution would have been to undo the change, but that was not deemed acceptable for some reason. As a result, the `no-exit` build option was introduced to save various platforms that did not behave exactly like `gcc`/`glibc`. While `atexit` is generally acceptable in a standalone project, OpenSSL is embedded in too many other projects and changes to `atexit` usage has negative impacts.

My opinion is that `atexit` has no business in OpenSSL given the behavioural variances in the `atexit` semantics when DLLs are involved. 3.3 or 4.0 should remove `atexit` completely and leave it up to the using projects instead, as an formal API change instead of what happened. This is just my opinion based on fighting this particular C capability for decades on various platform.

**Comment by kroeckx** [comment]
>So while it's the right behavior, and the best way for applications to fix it right now, OpenSSL still has something to do here.

I agree, but it seems unlikely that we change the behavior in a stable branch. I have no problem changing it for the next minor release (3.3).

**Comment by kroeckx** [comment]
> The use of `atexit` changed at 3.3 with new test introductions and changed existing behaviour of OpenSSL regardless of what was there "in the beginning". This broke some platforms as evidence from the number of `atexit` cases here. A potentially reasonable solution would have been to undo the change, but that was not deemed acceptable for some reason.

3.3 wasn't released yet. I assume you mean 3.2? 3.3 does have a new option to disable it at build time, that still seems to cause issues in some tests that seem to require atexit to be able to run the test.

As far as I know, nothing was changed regarding to the default use of atexit, it's been like this 1.1.0. Which change are you talking about?

But I think since the providers in 3.0 people run into more problems because of it.

**Comment by hyc** [comment]
> But I think since the providers in 3.0 people run into more problems because of it.

Sounds right. The crash that was reported to us only happens when the legacy provider is enabled.

**Comment by t8m** [comment]
**OTC: Would it be OK to make no-atexit default in 3.3 version?**

**Comment by rsbeckerca** [comment]
> > The use of `atexit` changed at 3.3 with new test introductions and changed existing behaviour of OpenSSL regardless of what was there "in the beginning". This broke some platforms as evidence from the number of `atexit` cases here. A potentially reasonable solution would have been to undo the change, but that was not deemed acceptable for some reason.
> 
> 3.3 wasn't released yet. I assume you mean 3.2? 3.3 does have a new option to disable it at build time, that still seems to cause issues in some tests that seem to require atexit to be able to run the test.
> 
> As far as I know, nothing was changed regarding to the default use of atexit, it's been like this 1.1.0. Which change are you talking about?
> 
> But I think since the providers in 3.0 people run into more problems because of it.

Actually I do mean 3.3. 3.2 worked broadly as far as I could tell. New tests in 3.3 caused SIGSEGV.

**Comment by hyc** [comment]
> Use a destructor when the compiler supports it; only use atexit when there's no support for destructors and you're building a static library. It's silly not to use a feature that everybody's compiler supports today, and choosing to rely on a nonstandard glibc behavior instead.

Unfortunately it's not a one-line fix since somewhere along the way your build process dropped the distinction in CFLAGS for static libs vs shared libs. Might have happened in 2016 or so, e.g. bbd9a50fa8e371927c290170ab6b8ac2597e5ff9

You go to the trouble of compiling separate object files for shared vs static libraries, but you use the shared_cflags for all of them so there's actually no difference in the objects. This also means you've lost any performance advantage that you would usually get with non-PIC static libraries. Very confusing.

**Comment by t8m** [comment]
**OTC: We will investigate making no-atexit default in 3.4 and use library destructors to call the OPENSSL_cleanup().**

**Comment by fwh-dc** [comment]
@nhorman Do you agree that this could be closed seeing as https://github.com/openssl/openssl/pull/29874 was merged
